### ISSUE TITLE
Keep old XL advancement for normal/hard

### ISSUE THREAD

**Issue by Sandman25DCSS** [jeremygurr/dcssca#203]
I've heard a player reached XL 20 after Lair/Orc/Snake 3 on hard without having any "experience" or other options, that makes the game too easy IMHO. I suggest to keep standard XL advancement (and max XL 27 too) for normal/hard mode.
For comparison, my Naga reached XL 20 in Vaults 4 after clearing both Lair branches

**Comment by Sandman25DCSS** [comment]
Oops, maybe my impression was wrong, the player did reach Swamp 4 too.

**Comment by jeremygurr** [comment]
Yeah that sounds about right to me. At least for normal mode. Maybe a bit much for hard mode still. The solution here though isn't to default to level 27 cap and old advancement, but to tweak the new to give a little less, which is both easier and more consistent. It won't be long before my fork is much longer and harder than vanilla crawl, and a level 27 cap just isn't going to cut it. It's lame to play 1/3 of the game without being able to advance your character.

After completing a 3 rune win in classic mode, a player of a given species should be at the same xl in normal difficulty as if they were playing vanilla crawl. That's my calibration point. What would you say that is? Maybe level 24? Hard mode should be a solid 3 or 4 levels lower, and easy probably 3 or 4 levels higher.

**Comment by Sandman25DCSS** [comment]
I don't see why we should have new XL advancement if it should be similar to old one. Can we just use old XL advancement instead? At least in classic mode for hard/normal?

**Comment by Sandman25DCSS** [comment]
"It's lame to play 1/3 of the game without being able to advance your character"

Sorry, I missed it initially. Are you talking about 15 rune game here? In normal crawl you usually get XL 27 soon after you obtain 3 runes, that's not 2/3 of the game, closer to 90%. And for 15 runes you still train skills even if HP/MP don't increase. XL 27 HP/MP is balanced/good enough for 15 runes since Torment exists. I wouldn't like Ogre having higher HP due to XL 28+, it is already very good.

**Comment by jeremygurr** [comment]
You can always use the level 27 cap option if you must, but for long term development this causes a problem. You are worried about the game becoming too easy for an ogre at level 35, which is understandable with the current situation. But the problem here isn't the level cap, it's the monster/game challenge cap. If the monsters are too easy for a level 35 ogre, I'd rather make things more challenging. The rune curses are a major example of this. Rune curses make it to where players and combos that could normally comfortably pull off a 15 rune win would be lucky to pull off a 5 rune win. So you take what would normally be quite overpowered, say a level 35 ogre, and maybe instead of a 5 rune win, with the same skill level, he could pull off a 7 rune win. Great, but there's still a lot of room for improvement. Yeah it will take a lot of work to get the rune curses balanced right, but it will get there. And there are several other changes that will similarly add to the extended game challenge. All together, there will be plenty of challenge even for a level 40 ogre. Thus the need for removing level caps. 

For any given build, at any achievable experience level, there should always be "tough spots" in the game. Maybe vaults is boring for one particular combo, but then gehenna is super challenging. But if there is any combo that has trouble finding a challenge anywhere, and can easily just sweep up runes, then there is a problem. It may be that the combo needs to be nerfed, but that's the quick and easy solution, and not always the right one. The harder, but often better solution is to add challenge to the game in such a way as to provide a dangerous place for that combo to go, without crippling other combos. 

Now this may be bringing out a more fundamental problem. Maybe characters shouldn't advance in health or magic at all between experience levels. It's already likely the way I will be implementing stamina. Maybe a species should start with fixed health and magic pools, which can only be increased with items or mutations. This can already be tried with the moon troll. You probably don't mind all that much if the ogre gains a few more stat points over the extended game, you are just concerned about the health pool going through the roof and removing any challenge there. This would be a huge rebalancing challenge of course, but I think one that would add great depth to the game and solve a major issue.

**Comment by Sandman25DCSS** [comment]
Currently there are no rune curses so removing XL limit does nothing but makes late game easier. Can you remove XL limit _after_ introducing rune curse?
Yesterday I started VSAs but suspended playing right on turn 1 after I realized that I will be XL 27 much earlier than I get 3 runes. My last win was Sp at XL 24, that was challenging and fun, it is no longer possible to have such low characters without voluntary limitations like speedruninng.
Also I disagree about nerfing easy combos, your crawl is great because it has difficulty modes and IMHO you should not try to balance combos. If some combo is too easy even in hard mode, just introduce "superhard" mode with -50% HP or something.

**Comment by jeremygurr** [comment]
I'm not sure what you mean about nerfing easy combos. I was trying to say that I didn't want to nerf any combo if I could possibly change the game in a way to keep them challenging. 

I'm okay with defaulting the level 27 cap to on for now. 

Having the cap off for normal 3 rune games should be just as difficult though. If it isn't, that just means that the experience curve is off, which is easily adjusted. Is that the case now? Do I need to adjust the non-capped experience curve to be 30% lower? I need real numbers so that it can be adjusted properly. Ideally if someone could do a 3 rune win and tell me what level they ended up with, and what level they would have expected, then that will help me tweak it.

**Comment by Sandman25DCSS** [comment]
I meant that I disagree with "The harder, but often better solution is to add challenge to the game in such a way as to provide a dangerous place for that combo to go, without crippling other combos". What you suggested here is a nerf of the easy combo (it gets harder, no other combos are affected) and I am suggesting to leave the combo unchanged and provide an option to make all combos harder instead.

Sorry, I didn't change my opinion on XL issue, I still believe playing in classic normal difficulty mode should be identical to standard crawl (XP-wise)

**Comment by jeremygurr** [comment]
what did you think about removing hp and mp increases between levels? That would greatly add to species distinction and simplify some late game balance issues. Hp and Mp could only be increased through items or mutations. Obviously characters would have to start with more than they have now. You seemed to really like the Moon Troll, which is mostly an application of this idea. It would require a fair bit of rebalancing, but I think the reward would be worth it.

**Comment by yrmvgh** [comment]
That honestly sounds like reinventing the wheel. You want to uncap levels, but then fix it by capping MP and HP. Although I guess I'm a fan in general of removing meaningless treadmills in RPGs (such as having players' and monsters' defense/accuracy all going up at the same rate as they level).

But if you're gonna remove MP/HP on level, at that point it almost becomes time to remove levels entirely, since they're just giving you stat points, spell slots, and AC (for some races).

**Comment by jeremygurr** [comment]
yeah levelling up is still meaningful for those reasons and for mutation advancement. There is a big difference between a game where your health and the monster's health and damage scale over most of the game, and then suddenly for no reason your health stops advancing, and a game where your health doesn't scale with experience at all. Although monsters get more difficult as you get deeper, they don't need to get as hard as before. 

Think of the difference in playing a moon troll vs. playing a cave troll. Even without recalibrating the monsters to better fit a non-hp scaling mechanic, you can feel a dramatic difference in game play. It's so much easier to balance late game and keep it interesting without the continuous health increases brought on by higher levels. An XL cap is a weak solution because it doesn't fully solve the balance issue, while at the same time creating a disappointing experience for the player. The player was motivated through most of the game by XL advancement leading to a more powerful character, but suddenly, just when the extended game is getting started, they can't advance any further. That deflates a lot of fun out of the late game.

**Comment by Sandman25DCSS** [comment]
I liked Moon Troll starting from middle game and later, basically at the stages where it had less HP than a normal Troll would have. I didn't like it on first floors, having 60-80 HP on D1 is boring, I would rather start game right in Lair with that HP.
I am glad you realize that crawl monsters are balanced around increasing HP, if you change it, you will have a completely different game like you do with pacifist mode. Sorry, I stopped playing the fork, it is not challenging with new XL advancement and it is hard to follow all those changes, I will wait for a stable version I guess.

**Comment by Sandman25DCSS** [comment]
I cannot say I agree with this: 

> The player was motivated through most of the game by XL advancement leading to a more powerful character, but suddenly, just when the extended game is getting started, they can't advance any further

Skill advancement is much more important, HP/MP advancement is just a nice  bonus. Even in extended while my character is still XL 27 it can be quite different after 3 runes, after 10 turns  and fter 15 runes. For example, I can get Necromutation, Statue Form, Death's Door, switch to Shadow Dragon Armour and Stealth, train ranged attack, change weapon category, change god etc.

**Comment by jeremygurr** [comment]
My point isn't that hp advancement is a bigger deal than skill training, I know that isn't the case. My point is that advancing your character through whatever way you can is the major driving force behind an RPG. If all advancement were to stop, the game would get lame very quick, no matter how many more monsters or branches were left to conquer. Although you still advance in vanilla crawl through being able to train other skills (those that haven't gotten to level 27 yet) and through finding new items (although after getting all resistances and the strongest weapons available, that wears out eventually too), it would be even better if you could continue to advance your XL, and advance your skills beyond level 27. 

Think of it this way: if the level 27 cap actually improved the game in a deep way (and wasn't just a superficial bandage over a deeper problem), then a lower cap would be even better. Why don't we cap all skills and XL at level 10? Or level 5? What makes 27 any better?

**Comment by jeremygurr** [comment]
Things will be quite unstable for a while as things are rebalanced. I'm thinking that maybe I should get more changes done at once, and then the rebalancing can happen all at once, instead of repeatedly rebalancing. The stamina change is getting close, but I'm thinking it would be a good idea to flatten hp and mp pools. You are absolutely right that we would need to adjust things a fair bit to make monsters challenging from the beginning, but I don't think that's very hard. It just takes some time. There are about 4 major features that would be the most balance breaking, and I think I will do them all at once. Then we can stabilize that and I can add some minor features and fix some bugs. After that is stable for a while then I will do another destabilizing round.

**Comment by Sandman25DCSS** [comment]
A few comments:
1) This week I got a game from another player (some crawl server was shutdown so server admin allowed to download the game) so I started playing right into 148th (yes, one hundred forty eight) ziggurat. Surprisingly I was able to upgrade my armour in a very significant way (got 22 extra AC for the cost of 5 EV, no spell failure rates changed, it was a +13 pearl dragon armour with Str/Int boosts) right in that ziggurat. Also I found some better rings but I felt lazy to use them. So we never know when progress stops.
2) XL 27 player is in extended where torment halves HP so even huge difference between 250 HP and 500 HP is not as significant as before, without rN+ you need just be tormented one more time. This is very easy to get, especially since there are vaults where you can open a door and see up to 7 monsters with torment.
3) I remember I created a thread on tavern forum complaining about extended game casting characters being too similar to each other, especially for ziggurat cleaners. They all have Haste, Deflect Missiles, Necromutation, great weapon at min delay, several level 9 spell etc. so I suggested to limit schools (for example, you cannot have more than 1 school above 25, more than 2 schools above 20 etc.) So I guess our opinions are quite different on the issue, I would like characters to stop improving earlier than in standard crawl while you seem to be trying to improve characters forever. Though I may be absolutely wrong here, you want to stop improving (no HP/MP growth) and have unlimited improving (no XL cap) at the same time.

**Comment by Sandman25DCSS** [comment]
XL 27 is better because this is what you get in 3 rune game. I often reach XL 27 right in Zot 5. So the cap purpose is to make extended hard, you basically need to get 12 more runes without HP/MP improvement.

**Comment by jeremygurr** [comment]
I get that the level 27 cap is meant to make extended hard, I just think there are better ways to do that. 

I want to take away one specific aspect of growth (hp can still grow through items, but other means would be removed), while eliminating xl caps. But even then the hp pool is not "capped" at any particular amount. If you want to focus all of your items on maxing out your hp, you can. And there's no reason that items couldn't get better and better throughout extended game. Likewise, although training and xl are not capped by a specific number, they are still effectively limited by the amount of experience a player can find in the game (which should never be unlimited in my opinion). So yes, I believe in limits, I just want to apply them from a different direction and see how that goes.

**Comment by Sandman25DCSS** [comment]
You sound like player always has items which increase HP... I know only one such item - amulet of vitality. And for the record I decided not to use it with my character who found it, spirit amulet was much better. I don't see why you want all extended characters be like current character of Chei and wear heavy armour while casting level 9 spells, this is what happens when you have unlimited bonus stats. Previously characters had interesting choice: do I want to have low Int as berserker for easy game but problems in extended, with unlimited stats you can do anything you like because it does not matter, you will still get Int 20 as TrBe before clearing Tomb. Also intrinsic AC+33 for Draconian/Naga and I don't even know how much AC for Gr will be "fun".

I agree, let's see what happens, just add rune curse and whatever you like _before_ removing cap and then you can decide how to solve issues which arise. I suspect no issues will need to be resolved, some players do crazy things.

**Comment by jeremygurr** [comment]
I don't mean to imply that the existing items would just automatically work with such a drastic change. I expect a fair bit of item tweaking would be necessary to make it work. For starters, a lot more items would need hp and mp enhancing buffs. 

For fun: level 27 naga has 9 base AC (checked after taking off all armour). Level 33 naga has 11 AC.

Level 27 gargoyle has 20 base AC. Level 33 gargoyle has 25. Armour skill doesn't seem to affect it. 

Yeah a 5 AC boost in the extreme case is something, but not exactly game breaking I don't think.

For fun I tried out a level 40 gargoyle. Base AC of 31. Which is definitely pretty cool. But considering the experience required to get to level 40 is over 6 times that which gets to level 27, that's just not going to happen, ever. Like I said earlier I don't believe in unlimited experience sources (because of my anti-grinding philosophy), and it won't be too long before I squash the ones that remain one way or another. I think half of the existing experience modes already put a hard cap on the amount of experience obtainable in a given game. 

As far as unlimited stats, that isn't how it would work. Experience costs per level still go up exponentially. A player could probably reach level 30 before too long in extended, but that is only 1 controlled stat point more than level 27 has (and possibly 1 species specific stat point). That's not really going to be a game changer (assuming the hp and mp aren't changing). 

The amount of experience that it takes to get to level 27 is 332k. The amount of experience to get from level 27 to level 33 (and thus get 2 more stat points + probably 1 species point) is 409k, which is significantly more than all of the experience that the level 27 character had gotten in their entire game so far. That's not happening any time soon, even in extended. And even if it did, we are talking 3 total stat points, and possibly a few more ac. 

I think reaching level 27 at Zot 5 with 3 runes on normal mode, with exp apt of 0, is a perfect calibration point for all experience modes. If I discover that any of the existing experience modes are off of this standard, then I will adjust them until they are. Hard mode, with exp apt of -1, should be closer to level 23 by Zot 5.

**Comment by Sandman25DCSS** [comment]
My AC numbers were for XL 99 character.
I guess I should stop arguing at this point, I am completely lost. Max XL 99 but only 30 is really reachable, no HP/MP increase but adding extra items with HP/MP (poor characters who are unlucky). Do whatever you want, you are creating a new game of different genre. Thank you, I enjoyed DCSS CA while it lasted ;)

**Comment by jeremygurr** [comment]
default is now level_27_cap = true (applies to the next release). I have put off further experience changes to be done later, when many other experience rebalancing issues will be resolved at the same time. For now I'm focussing on the stamina change and getting the rune curses done.
